kmip/services/server/engine.py

```diff
@@ -0,0 +1,404 @@
+                    )
+                    raise exceptions.OperationNotSupported(
+                        "{0} is not supported by KMIP {1}".format(
+                            operation,
+                            self._protocol_version
+                        )
+                    )
+                else:
+                    return function(self, *args, **kwargs)
+            return wrapper
+        return decorator
+
+    def _synchronize(function):
+        def decorator(self, *args, **kwargs):
+            with self._lock:
+                return function(self, *args, **kwargs)
+        return decorator
+
+    def _set_protocol_version(self, protocol_version):
+        if protocol_version in self._protocol_versions:
+            self._protocol_version = protocol_version
+        else:
+            raise exceptions.InvalidMessage(
+                "KMIP {0} is not supported by the server.".format(
+                    protocol_version
+                )
+            )
+
+    def _verify_credential(self, request_credential, connection_credential):
+        # TODO (peterhamilton) Add authentication support
+        # 1. If present, verify user ID of connection_credential is valid user.
+        # 2. If present, verify request_credential is valid credential.
+        # 3. If both present, verify that they are compliant with each other.
+        # 4. If neither present, set server to only allow Query operations.
+        pass
+
+    @_synchronize
+    def process_request(self, request, credential=None):
+        """
+        Process a KMIP request message.
+
+        This routine is the main driver of the KmipEngine. It breaks apart and
+        processes the request header, handles any message errors that may
+        result, and then passes the set of request batch items on for
+        processing. This routine is thread-safe, allowing multiple client
+        connections to use the same KmipEngine.
+
+        Args:
+            request (RequestMessage): The request message containing the batch
+                items to be processed.
+            credential (Credential): A credential containing any identifying
+                information about the client obtained from the client
+                certificate. Optional, defaults to None.
+
+        Returns:
+            ResponseMessage: The response containing all of the results from
+                the request batch items.
+        """
+        header = request.request_header
+
+        # Process the protocol version
+        self._set_protocol_version(header.protocol_version)
+
+        # Process the maximum response size
+        max_response_size = None
+        if header.maximum_response_size:
+            max_response_size = header.maximum_response_size.value
+
+        # Process the time stamp
+        if header.time_stamp:
+            then = header.time_stamp.value
+            now = int(time.time())
+
+            if (now >= then) and ((now - then) < 60):
```

Comment:
Ah, I see. Then the log message should state that invalid time. It could be past or future time.